### PR TITLE
misc: Add browser support check and update supported browser messages

### LINKED ISSUE #28914: Better message when a supported browser in not found.
### What would you like?

`Browser: chrome was not found on your system or is not supported by Cypress.`

Should be 

`Browser: chrome was not found on your system.`

### Why is this needed?

If any of the supported browsers is not found, a specific message would be better. This is a better user experience.

If a non-supported browser is specified, e.g. `abc`, then `Browser: abc is not supported by Cypress.` is also a little better.

### Other

_No response_

### PR DESCRIPTION
**Pull request by 0v00**
- Closes [https://github.com/cypress-io/cypress/issues/28914](https://github.com/cypress-io/cypress/issues/28914)

### Additional details

Current messages for browser support/browser's not found can be confusing e.g.:
`Browser: chrome was not found on your system or is not supported by Cypress.`
This PR aims to provide clear messaging if a browser is not supported, or if it was not found by name or path.
- If a user enters a browser name or path that does not contain any supported browser names, or does not contain any custom browser (e.g. Brave) names added via Cypress config, then the user will see a `BROWSER_NOT_SUPPORTED` error message.
- I have added a regex that should cast a wide net for supported browsers - we ideally want to have zero false negatives. If a user enters a browser name or path that is supported (e.g. `C:\some\path\msedge.exe`) but is not found on their system, they will still see the appropriate `BROWSER_NOT_FOUND_BY_NAME` or `BROWSER_NOT_FOUND_BY_PATH` error. However, if a user misspells a browser name e.g. `webkitt`, or enters an unsupported browser name e.g. `arc`, they will see the `BROWSER_NOT_SUPPORTED` error message.
- I have moved the special `canary` error message to the `BROWSER_NOT_SUPPORTED` error message.
- `getBrowserLauncher()` now throws the `BROWSER_NOT_SUPPORTED` error instead of `BROWSER_NOT_FOUND_BY_NAME` - if `electron` or one of the supported browser families is not found, then it should be safe to assume that this browser is not supported.
- I have added `webkit` to the list of supported browsers in the `BROWSER_NOT_SUPPORTED` error message.
- Custom browsers added via Cypress config still work and are not filtered out.
- I have updated the `BROWSER_NOT_FOUND_BY_NAME` error - this error message may still need to be made clearer.
- This PR doesn't take channels into account e.g. passing in `chrome:someunsupportedchannel` - this will throw a `BROWSER_NOT_FOUND_BY_NAME` error. I think this behavior makes sense as we want to avoid any false negatives - if the name/path contains e.g. `chrome`, we can assume it's supported and then throw a more appropriate error later on.
- _*(Apologies for the number of commits)*_

### Steps to test

Test the following commands:
- `yarn cypress:run --browser unsupportedbrowser --project /your/project`
- `yarn cypress:run --browser badname:channel--project /your/project`
- `yarn cypress:run --browser /path/to/unsupported/browser --project /your/project`
- Run `yarn test test/unit/browsers/browsers_spec.js` to verify passing tests
- Run `yarn test` in `packages/errors` to verify passing tests

### How has the user experience changed?

- If a user enters a browser name or path that does not match any supported browser names, or does not match any custom browsers (e.g. Brave) added in via Cypress config, then the user will see a 'browser not supported' error message.
unsupported_path
unsupported_name
unsupported_name_channel

### PR Tasks

- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by 0v00** [comment]
@jennifer-shehane, recreated this PR as the last one had a messy git history/I made some mistakes when rebasing/merging - apologies for any confusion/extra work on your end.

Thank you for sending @emilyrohrbough 's PR (https://github.com/cypress-io/cypress/pull/28377/files). A few thoughts around adding `webkit` and `electron` directly to the `knownBrowsers` object:
- Does the Electron browser have a binary or a versionRegex? If not, I wonder if it might be better to export it separately in `known-browsers` with its own type defined in `types/src/browsers.ts` e.g. `export type ElectronBrowser = Omit<Browser, 'versionRegex' | 'binary'>`. Otherwise we could make the `binary` and `versionRegex` properties in `Browser` optional and add Electron to `knownBrowsers`, this would require a few changes in `launcher/lib/detect.ts` (not sure what else we would have to change).

- The Webkit browser in @emilyrohrbough draft PR conforms to the `Browser` type, but adding it directly to the `knownBrowsers` object will require some changes to `launcher/lib/windows/index.ts` and `launcher/lib/darwin/index.ts` in order to get the `darwin_spec.ts` and `windows_spec.ts` tests passing. Same for Electron.

- Adding these both directly to the `knownBrowsers` I believe will require some additional changes to browser detection in `server/lib/browsers/utils.ts`, specifically in `getBrowsers()`

I think the potential options are: 
- add both `webkit` and `electron` to `known-browsers.ts` but exported separately from the `knownBrowsers` object (and look at properly integrating webkit and electron into knownBrowsers in another PR)
- add them into the `knownBrowsers` object in this PR (but this PR will get a bit bigger in scope)

I'm not sure if there is a better solution that I'm missing entirely (I could just be overthinking this).

File: packages/errors/src/errors.ts [hunk 1]
```diff
@@ -102,30 +102,42 @@ export const AllCypressErrors = {
...
     return errTemplate`\
-        Can't run because you've entered an invalid browser name.
+        Can't run because you've entered a browser that is not supported by Cypress.
 
-        Browser: ${fmt.highlight(browser)} was not found on your system or is not supported by Cypress.
+        Browser: ${fmt.highlight(browser)} is not supported by Cypress.
 
         Cypress supports the following browsers:
-        ${fmt.listItems(['electron', 'chrome', 'chromium', 'chrome:canary', 'edge', 'firefox'])}
+        ${fmt.listItems(['electron', 'chrome', 'chromium', 'chrome:canary', 'edge', 'firefox', 'webkit'])}
```
Comment (cacieprins) [inline packages/errors/src/errors.ts:123]: Should we note here that webkit support is still experimental?

**Review by cacieprins — CHANGES_REQUESTED** [review]
Thank you for this PR! I think this is a great start to improving the developer experience with these error messages.

I do think that this change is a great time to combine webkit and electron into the KnownBrowsers list, though, despite the additional scope.

File: packages/errors/src/errors.ts (hunk 1 above)
Comment (0v00) [reply packages/errors/src/errors.ts:123]: @cacieprins @jennifer-shehane sorry for the delay, will jump back on the pr this week

**Review by AtofStryker — CHANGES_REQUESTED** [review]
Looks like a few system tests need to be updated? https://app.circleci.com/pipelines/github/cypress-io/cypress/60732/workflows/ada99278-62c9-4dbd-bbf1-e49ef1444d99/jobs/2546978/tests#failed-test-2

@0v00 let us know if you need help updating those!

**Comment by jennifer-shehane** [comment]
@0v00 Will you be able to address the failing system tests?

**Comment by 0v00** [comment]
> @0v00 Will you be able to address the failing system tests?

yes, i think i'll have to update some of the unit tests in the launcher package as well.

**Comment by 0v00** [comment]
@AtofStryker @cacieprins I have added `webkit` and `electron` into the `knownBrowsers` list (in `known-browsers.ts`), and now the `detects browsers as expected` unit test in `windows_spec.ts` (`launcher/test/unit/windows_spec.ts`) will need to be updated. Do either of you have any thoughts on the best way to stub out the webkit and electron browsers for this test, and more specifically, where electron and webkit are actually installed on windows?

**Comment by cacieprins** [comment]
I think it's probably fairly safe to just strip Electron and Webkit from knownBrowsers for windows in the tests: electron is baked into the Cypress binary, and Webkit isn't available for windows. For linux, both can be stripped as well.

**Comment by 0v00** [comment]
> I think it's probably fairly safe to just strip Electron and Webkit from knownBrowsers for windows in the tests: electron is baked into the Cypress binary, and Webkit isn't available for windows. For linux, both can be stripped as well.

@cacieprins i think we should be good to try rerunning the tests now

**Comment by jennifer-shehane** [comment]
@0v00 Could you add a changelog entry? https://github.com/cypress-io/cypress/actions/runs/9460366383/job/26059048449?pr=29066

**Comment by jennifer-shehane** [comment]
@0v00 Also there look to be some failures due to these changes if you can investigate.

**Comment by jennifer-shehane** [comment]
@0v00 We'll be closing this PR if you cannot address the remaining work.

**Comment by 0v00** [comment]
> @0v00 We'll be closing this PR if you cannot address the remaining work.

Remaining work has been addressed and tests are all passing, but re: the CHANGELOG.md, what is the next release date and version number to add in order to pass the Semantic Pull Request check?

Current unit-tests failure seems unrelated to this PR:
```bash
  1 failing

  1) Proxy generating certificates handles errors with reusing existing certificates:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
      at listOnTimeout (node:internal/timers:569:17)
      at processTimers (node:internal/timers:512:7)
```

**Comment by MikeMcC399** [comment]
- Will this PR resolve issue https://github.com/cypress-io/cypress/issues/27304?

> I think it's probably fairly safe to just strip Electron and Webkit from knownBrowsers for windows in the tests: electron is baked into the Cypress binary, and Webkit isn't available for windows. For linux, both can be stripped as well.

- I didn't understand the reference to WebKit not being available in Windows (see https://github.com/cypress-io/cypress/pull/29066#issuecomment-2150368275). The repro for issue #27304 shows WebKit running on Windows:

```text
  (Run Starting)
  ┌────────────────────────────────────────────────────────────────────────────────────────────
  │ Cypress:        13.13.0                                                                        │
  │ Browser:        WebKit 17 (headless)                                                           │
  │ Node Version:   v20.13.1 (c:\runners\2.317.0\externals\node20\bin\node.exe)                    │
  │ Specs:          1 found (spec.cy.js)                                                           │
  │ Searched:       cypress/e2e/**/*.cy.{js,jsx,ts,tsx}                                            │
  │ Experiments:    experimentalWebKitSupport=true                                                 │
  └────────────────────────────────────────────────────────────────────────────────────────────
─────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  spec.cy.js                                                                      (1 of 1)
before launching browser
{
  name: 'webkit',
  channel: 'stable',
  family: 'webkit',
  displayName: 'WebKit',
  version: '17.4',
  path: 'C:\\Users\\runneradmin\\AppData\\Local\\ms-playwright\\webkit-2035\\Playwright.exe',
  majorVersion: '17',
  warning: 'WebKit support is currently experimental. Some functions may not work as expected.',
  isHeadless: true,
  isHeaded: false
}
```

File: packages/launcher/test/unit/windows_spec.ts [hunk 2]
```diff
@@ -35,6 +35,13 @@ function detect (goalBrowsers: Browser[]) {
 
 const HOMEDIR = 'C:/Users/flotwig'
 
+// Electron and WebKit can be safely removed since Electron is baked into the Cypress binary
```
Comment (cacieprins) [inline packages/launcher/test/unit/windows_spec.ts:38]: I was wrong about Webkit - I'm very sorry! Thank you @MikeMcC399 for bringing this up - I was going off of what is generally available on Windows, and did not realize we were recommending the use of playwright-webkit for windows support.

**Review by cacieprins — CHANGES_REQUESTED** [review]

**Comment by jennifer-shehane** [comment]
@0v00 Will you be able to address the feedback?

**Pull request closed without merging**